File: examples/python/quickstart.py
"""NexaLayer Python quickstart.

Creates a dynamic session, reports telemetry, reads health, and always
terminates the session in cleanup. This can create a real paid session.
"""

import os
import time
import uuid

import requests
from nexalayer import NexaLayerClient


def unwrap(response):
    return response.get("data", response)


def pick_dynamic_product(client: NexaLayerClient) -> str:
    override = os.getenv("NEXALAYER_PRODUCT_NO")
    if override:
        return override
    products = unwrap(client.get_products(type="dynamic"))
    for item in products.get("items", []):
        if item.get("product_no"):
            return item["product_no"]
    raise RuntimeError("No dynamic product available")


def wait_for_active(client: NexaLayerClient, session_id: str) -> dict:
    for _ in range(60):
        session = unwrap(client.get_session(session_id))
        if session.get("status") == "active":
            return session
        if session.get("status") in {"failed", "error"}:
            raise RuntimeError(f"Session failed: {session}")
        time.sleep(2)
    raise TimeoutError("Session did not become active within 120s")


def main():
    api_key = os.getenv("NEXALAYER_API_KEY")
    if not api_key:
        raise RuntimeError("Set NEXALAYER_API_KEY")

    client = NexaLayerClient(
        api_key=api_key,
        base_url=os.getenv("NEXALAYER_BASE_URL", "https://api.nexalayer.net/v1"),
    )
    session_id = None

    try:
        product_no = pick_dynamic_product(client)
        session = client.create_session(
            session_type="dynamic",
            product_no=product_no,
            quantity=1,
            protocol="socks5",
            rotation_mode="on_demand",
            idempotency_key=f"sdk-py-{uuid.uuid4()}",
        )
        session_id = session.session_id
        print("session_id:", session_id)

        active = wait_for_active(client, session_id)
        proxy_url = active.get("proxy", {}).get("full_url")
        if not proxy_url:
            raise RuntimeError("Active session did not include proxy.full_url")

        try:
            res = requests.get(
                "https://httpbin.org/ip",
                proxies={"http": proxy_url, "https": proxy_url},
                timeout=30,
            )
            event_type = "success" if res.ok else "http_error"
            status_code = res.status_code
            print("httpbin:", res.text[:120])
        except requests.Timeout:
            event_type = "timeout"
            status_code = 0

        client.report_event(
            session_id,
            event_type=event_type,
            status_code=status_code,
            target_host="httpbin.org",
        )
        print("health:", client.get_session_health(session_id))
    finally:
        if session_id:
            client.terminate_session(session_id)
            print("terminated:", session_id)


if __name__ == "__main__":
    main()

File: examples/node/quickstart.ts
/**
 * NexaLayer Node quickstart.
 *
 * Creates a dynamic session, reports telemetry, reads health, and always
 * terminates the session in cleanup. This can create a real paid session.
 */

import { NexaLayerClient } from '../../node/src';

function unwrap(value: unknown): Record<string, unknown> {
  const obj = value as Record<string, unknown>;
  return (obj.data as Record<string, unknown>) ?? obj;
}

async function pickDynamicProduct(client: NexaLayerClient): Promise<string> {
  if (process.env.NEXALAYER_PRODUCT_NO) {
    return process.env.NEXALAYER_PRODUCT_NO;
  }
  const products = unwrap(await client.getProducts({ type: 'dynamic' }));
  const items = (products.items as Array<Record<string, unknown>>) ?? [];
  const product = items.find((item) => item.product_no);
  if (!product?.product_no) {
    throw new Error('No dynamic product available');
  }
  return String(product.product_no);
}

async function waitForActive(
  client: NexaLayerClient,
  sessionId: string
): Promise<Record<string, unknown>> {
  for (let i = 0; i < 60; i += 1) {
    const session = unwrap(await client.getSession(sessionId));
    if (session.status === 'active') return session;
    if (session.status === 'failed' || session.status === 'error') {
      throw new Error(`Session failed: ${JSON.stringify(session)}`);
    }
    await new Promise((resolve) => setTimeout(resolve, 2000));
  }
  throw new Error('Session did not become active within 120s');
}

async function main() {
  const apiKey = process.env.NEXALAYER_API_KEY;
  if (!apiKey) throw new Error('Set NEXALAYER_API_KEY');

  const client = new NexaLayerClient({
    apiKey,
    baseUrl: process.env.NEXALAYER_BASE_URL ?? 'https://api.nexalayer.net/v1',
  });
  let sessionId: string | undefined;

  try {
    const productNo = await pickDynamicProduct(client);
    const session = await client.createSession({
      session_type: 'dynamic',
      product_no: productNo,
      quantity: 1,
      protocol: 'socks5',
      rotation_mode: 'on_demand',
      idempotencyKey: `sdk-node-${Date.now()}`,
    });
    sessionId = session.sessionId;
    console.log('session_id:', sessionId);

    const active = await waitForActive(client, sessionId);
    const proxy = active.proxy as Record<string, unknown> | undefined;
    if (!proxy?.full_url) {
      throw new Error('Active session did not include proxy.full_url');
    }

    await client.reportEvent(sessionId, {
      event_type: 'success',
      status_code: 200,
      target_host: 'httpbin.org',
    });
    console.log('health:', await client.getSessionHealth(sessionId));
  } finally {
    if (sessionId) {
      await client.terminateSession(sessionId);
      console.log('terminated:', sessionId);
    }
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
